File: src/tables/cmap.cpp
#include "src/algo.h"
#include "name.h"
#include "tables.h"

using namespace Name;

static void parseLanguage16(const PlatformID platformID, Parser &parser)
{
    const auto id = parser.peek<UInt16>();
    parser.readValue<UInt16>("Language ID", languageName(platformID, id));
}

static void parseLanguage32(const PlatformID platformID, Parser &parser)
{
    const auto id = parser.peek<UInt32>();
    parser.readValue<UInt32>("Language ID", languageName(platformID, quint16(id)));
}

static void parseFormat0(const PlatformID platformID, Parser &parser)
{
    parser.read<UInt16>("Subtable size");
    parseLanguage16(platformID, parser);
    parser.readBasicArray<UInt8>("Glyphs", 256);
}

static void parseFormat2(const PlatformID platformID, Parser &parser)
{
    const auto tableStart = parser.offset() - 2;
    const auto tableSize = parser.read<UInt16>("Subtable size");
    parseLanguage16(platformID, parser);

    quint16 subHeadersCount = 0;
    parser.readArray("SubHeader Keys", 256, [&](const auto index){
        const quint16 key = parser.read<UInt16>(index);
        subHeadersCount = qMax(subHeadersCount, quint16(key / 8));
    });

    parser.readArray("SubHeader Records", subHeadersCount + 1, [&](const auto index){
        parser.beginGroup(index);
        parser.read<UInt16>("First valid low byte");
        parser.read<UInt16>("Number of valid low bytes");
        parser.read<Int16>("ID delta");
        parser.read<UInt16>("ID range offset");
        parser.endGroup();
    });

    // TODO: technically, we should split the tail into subarrays,
    //       but looks like ranges can overlap and ttf-explorer doesn't support this

    const auto tailSize = tableSize - (parser.offset() - tableStart);
    parser.readBasicArray<GlyphId>("Glyph index array", tailSize / 2);
}

static void parseFormat4(const PlatformID platformID, Parser &parser)
{
    const auto tableStart = parser.offset() - 2;
    const auto tableSize = parser.read<UInt16>("Subtable size");
    parseLanguage16(platformID, parser);
    const auto segCount2 = parser.read<UInt16>("2 × segCount");
    const auto segCount = quint16(segCount2 / 2);
    parser.read<UInt16>("Search range");
    parser.read<UInt16>("Entry selector");
    parser.read<UInt16>("Range shift");
    parser.readBasicArray<UInt16>("End Character Codes", segCount);
    parser.read<UInt16>("Reserved");
    parser.readBasicArray<UInt16>("Start Character Codes", segCount);
    parser.readBasicArray<Int16>("Deltas", segCount);
    parser.readBasicArray<UInt16>("Offsets into Glyph Index Array", segCount);

    const auto tailSize = tableSize - (parser.offset() - tableStart);
    parser.readBasicArray<GlyphId>("Glyph Index Array", tailSize / 2);
}

static void parseFormat6(const PlatformID platformID, Parser &parser)
{
    parser.read<UInt16>("Subtable size");
    parseLanguage16(platformID, parser);
    parser.read<UInt16>("First code");
    const auto count = parser.read<UInt16>("Number of codes");
    parser.readBasicArray<GlyphId>("Glyph Index Array", count);
}

static void parseFormat8(const PlatformID platformID, Parser &parser)
{
    parser.read<UInt16>("Reserved");
    parser.read<UInt32>("Subtable size");
    parseLanguage32(platformID, parser);
    parser.readBytes("Packed data", 8192);
    const auto count = parser.read<UInt32>("Number of groups");

    parser.readArray("SequentialMapGroup Records", count, [&](const auto index){
        parser.beginGroup(index);
        parser.read<UInt32>("First character code");
        parser.read<UInt32>("Last character code");
        parser.read<UInt32>("Starting glyph index");
        parser.endGroup();
    });
}

static void parseFormat10(const PlatformID platformID, Parser &parser)
{
    parser.read<UInt16>("Reserved");
    parser.read<UInt32>("Subtable size");
    parseLanguage32(platformID, parser);
    parser.read<UInt32>("First code");
    const auto count = parser.read<UInt32>("Number of codes");
    parser.readBasicArray<GlyphId>("Glyph Index Array", count);
}

static void parseFormat12(const PlatformID platformID, Parser &parser)
{
    parser.read<UInt16>("Reserved");
    parser.read<UInt32>("Subtable size");
    parseLanguage32(platformID, parser);
    const auto count = parser.read<UInt32>("Number of groups");

    parser.readArray("SequentialMapGroup Records", count, [&](const auto index){
        parser.beginGroup(index);
        parser.read<UInt32>("First character code");
        parser.read<UInt32>("Last character code");
        parser.read<UInt32>("Starting glyph index");
        parser.endGroup();
    });
}

static void parseFormat13(const PlatformID platformID, Parser &parser)
{
    parser.read<UInt16>("Reserved");
    parser.read<UInt32>("Subtable size");
    parseLanguage32(platformID, parser);
    const auto count = parser.read<UInt32>("Number of groups");

    parser.readArray("ConstantMapGroup Records", count, [&](const auto index){
        parser.beginGroup(index);
        parser.read<UInt32>("First character code");
        parser.read<UInt32>("Last character code");
        parser.read<UInt32>("Glyph index");
        parser.endGroup();
    });
}

static void parseFormat14(Parser &parser)
{
    const auto tableStart = parser.offset() - 2;

    parser.read<UInt32>("Subtable size");
    const auto count = parser.read<UInt32>("Number of records");

    struct Record
    {
        bool isDefault;
        quint32 offset;
    };

    QVector<Record> records;
    parser.readArray("VariationSelector Records", count, [&](const auto index){
        parser.beginGroup(index);
        parser.read<UInt24>("Variation selector");
        const auto defOffset = parser.read<Offset32>("Offset to Default UVS Table");
        const auto nonDefOffset = parser.read<Offset32>("Offset to Non-Default UVS Table");
        parser.endGroup();

        if (defOffset != 0) {
            records.append({ true, tableStart + defOffset });
        }

        if (nonDefOffset != 0) {
            records.append({ false, tableStart + nonDefOffset });
        }
    });

    algo::sort_all_by_key(records, &Record::offset);
    algo::dedup_vector_by_key(records, &Record::offset);

    for (const auto record : records) {
        parser.advanceTo(record.offset);
        if (record.isDefault) {
            parser.beginGroup("Default UVS table");
            const auto count = parser.read<UInt32>("Number of Unicode character ranges");
            for (uint i = 0; i < count; ++i) {
                parser.beginGroup("Unicode range");
                parser.read<UInt24>("First value in this range");
                parser.read<UInt8>("Number of additional values");
                parser.endGroup();
            }
            parser.endGroup();
        } else {
            parser.beginGroup("Non-Default UVS table");
            const auto count = parser.read<UInt32>("Number of UVS Mappings");
            for (uint i = 0; i < count; ++i) {
                parser.beginGroup("UVS mapping");
                parser.read<UInt24>("Base Unicode value");
                parser.read<GlyphId>("Glyph ID");
                parser.endGroup();
            }
            parser.endGroup();
        }
    }
}

void parseCmap(Parser &parser)
{
    const auto tableStart = parser.offset();

    const auto version = parser.read<UInt16>("Version");
    if (version != 0) {
        throw QString("invalid table version");
    }

    struct Record
    {
        quint32 offset;
        PlatformID platformID;
    };

    const auto numberOfTables = parser.read<UInt16>("Number of tables");
    QVector<Record> records;

    parser.readArray("Encoding Records", numberOfTables, [&](const auto index){
        parser.beginGroup(index);
        const auto platformID = parser.read<PlatformID>("Platform ID");
        {
            const auto id = parser.peek<UInt16>();
            parser.readValue<UInt16>("Encoding ID", encodingName(platformID, id));
        }
        const auto offset = parser.read<Offset32>("Offset");
        parser.endGroup();

        records.append({ offset, platformID });
    });

    algo::sort_all_by_key(records, &Record::offset);
    algo::dedup_vector_by_key(records, &Record::offset);

    for (const auto record : records) {
        parser.advanceTo(tableStart + record.offset);
        parser.beginGroup();
        QString title;
        const auto format = parser.read<UInt16>("Format");
        switch (format) {
            case 0 : {
                parseFormat0(record.platformID, parser);
                title = "Byte encoding table";
                break;
            }
            case 2 : {
                parseFormat2(record.platformID, parser);
                title = "High-byte mapping through table";
                break;
            }
            case 4 : {
                parseFormat4(record.platformID, parser);
                title = "Segment mapping to delta values";
                break;
            }
            case 6 : {
                parseFormat6(record.platformID, parser);
                title = "Trimmed table mapping";
                break;
            }
            case 8 : {
                parseFormat8(record.platformID, parser);
                title = "Mixed 16-bit and 32-bit coverage";
                break;
            }
            case 10 : {
                parseFormat10(record.platformID, parser);
                title = "Trimmed array";
                break;
            }
            case 12 : {
                parseFormat12(record.platformID, parser);
                title = "Segmented coverage";
                break;
            }
            case 13 : {
                parseFormat13(record.platformID, parser);
                title = "Many-to-one range mappings";
                break;
            }
            case 14 : {
                parseFormat14(parser);
                title = "Unicode variation sequences";
                break;
            }
            default : break;
        }

        parser.endGroup(QString("Subtable %1").arg(format), title);
    }
}

File: src/tables/cmap.rs
use crate::parser::*;
use crate::{TitleKind, Error, Result};
use super::name::{self, PlatformId};

struct Record {
    offset: usize,
    platform_id: PlatformId,
}

pub fn parse(parser: &mut Parser) -> Result<()> {
    let table_start = parser.offset();

    let version = parser.read::<u16>("Version")?;
    if version != 0 {
        return Err(Error::InvalidTableVersion);
    }

    let num_tables = parser.read::<u16>("Number of tables")?;
    if num_tables == 0 {
        return Ok(());
    }

    let mut records = Vec::new();
    parser.begin_group("Encoding records");
    for i in 0..num_tables {
        parser.begin_group(format!("Record {}", i));
        let platform_id = parser.read::<PlatformId>("Platform ID")?;
        {
            let id = parser.peek::<u16>()?;
            parser.read_value(2, "Encoding ID", name::encoding_name(platform_id, id))?;
        }
        let offset = parser.read::<Offset32>("Offset")?.to_usize();
        parser.end_group();

        records.push(Record { offset, platform_id });
    }
    parser.end_group();

    records.sort_by_key(|r| r.offset);
    records.dedup_by_key(|r| r.offset);

    for record in records {
        parser.jump_to(table_start + record.offset)?;
        parser.begin_group("Table");
        let format = parser.read::<u16>("Format")?;
        let title = match format {
            0 => {
                parse_format0(record.platform_id, parser)?;
                "Byte encoding table"
            }
            2 => {
                parse_format2(record.platform_id, parser)?;
                "High-byte mapping through table"
            }
            4 => {
                parse_format4(record.platform_id, parser)?;
                "Segment mapping to delta values"
            }
            6 => {
                parse_format6(record.platform_id, parser)?;
                "Trimmed table mapping"
            }
            8 => {
                parse_format8(record.platform_id, parser)?;
                "Mixed 16-bit and 32-bit coverage"
            }
            10 => {
                parse_format10(record.platform_id, parser)?;
                "Trimmed array"
            }
            12 => {
                parse_format12(record.platform_id, parser)?;
                "Segmented coverage"
            }
            13 => {
                parse_format13(record.platform_id, parser)?;
                "Many-to-one range mappings"
            }
            14 => {
                parse_format14(parser)?;
                "Unicode Variation Sequences"
            }
            _ => {
                return Err(Error::Custom(format!("{} is not a valid subtable format", format)));
            }
        };

        parser.end_group_with_title(format!("Subtable {}: {}", format, title));
    }

    Ok(())
}

fn parse_language16(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    let id = parser.peek::<u16>()?;
    parser.read_value(2, "Language ID", name::language_name(platform_id, id))
}

fn parse_language32(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    let id = parser.peek::<u32>()?;
    parser.read_value(4, "Language ID", name::language_name(platform_id, id as u16)) // TODO: check cast
}

fn parse_format0(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    parser.read::<u16>("Subtable size")?;
    parse_language16(platform_id, parser)?;
    parser.read_array::<u8>("Glyphs", TitleKind::Glyph, 256)
}

fn parse_format2(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    let table_start = parser.offset() - 2;
    let table_size = parser.read::<u16>("Subtable size")? as usize;
    parse_language16(platform_id, parser)?;

    let mut sub_headers_count = 0;
    parser.begin_group_with_value("SubHeader keys", "256");
    for i in 0..256 {
        let key = parser.read2::<u16>(format!("Key {}", i))?; // TODO: this
        sub_headers_count = std::cmp::max(sub_headers_count, (key / 8) as u16);
    }
    parser.end_group();

    sub_headers_count += 1;
    parser.begin_group_with_value("SubHeader records", sub_headers_count.to_string());
    for i in 0..sub_headers_count {
        parser.begin_group(format!("SubHeader {}", i));
        parser.read::<u16>("First valid low byte")?;
        parser.read::<u16>("Number of valid low bytes")?;
        parser.read::<i16>("ID delta")?;
        parser.read::<u16>("ID range offset")?;
        parser.end_group();
    }
    parser.end_group();

    // TODO: technically, we should split the tail into subarrays,
    //       but looks like ranges can overlap and ttf-explorer doesn't support this

    let tail_size = table_size - (parser.offset() - table_start);
    parser.read_array::<GlyphId>("Glyph index array", TitleKind::Glyph, tail_size / 2)?;

    Ok(())
}

fn parse_format4(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    let table_start = parser.offset() - 2;
    let table_size = parser.read::<u16>("Subtable size")? as usize;
    parse_language16(platform_id, parser)?;
    let seg_count2 = parser.read::<u16>("2 × segCount")?;
    let seg_count = (seg_count2 / 2) as usize;
    parser.read::<u16>("Search range")?;
    parser.read::<u16>("Entry selector")?;
    parser.read::<u16>("Range shift")?;
    parser.read_array::<u16>("End character codes", TitleKind::Code, seg_count)?;
    parser.read::<u16>("Reserved")?;
    parser.read_array::<u16>("Start character codes", TitleKind::Code, seg_count)?;
    parser.read_array::<i16>("Deltas", TitleKind::Delta, seg_count)?;
    parser.read_array::<u16>("Offsets into Glyph index array", TitleKind::Offset, seg_count)?;

    let tail_size = table_size - (parser.offset() - table_start);
    parser.read_array::<GlyphId>("Glyph index array", TitleKind::Glyph, tail_size / 2)?;

    Ok(())
}

fn parse_format6(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    parser.read::<u16>("Subtable size")?;
    parse_language16(platform_id, parser)?;
    parser.read::<u16>("First code")?;
    let count = parser.read::<u16>("Number of codes")? as usize;
    parser.read_array::<GlyphId>("Glyph index array", TitleKind::Glyph, count)?;

    Ok(())
}

fn parse_format8(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    parser.read::<u16>("Reserved")?;
    parser.read::<u32>("Subtable size")?;
    parse_language32(platform_id, parser)?;
    parser.read_bytes(8192, "Packed data")?;
    let count = parser.read::<u32>("Number of groups")?;

    parser.begin_group_with_value("SequentialMapGroup records", count.to_string());
    for i in 0..count {
        parser.begin_group(format!("Record {}", i));
        parser.read::<u32>("First character code")?;
        parser.read::<u32>("Last character code")?;
        parser.read::<u32>("Starting glyph index")?;
        parser.end_group();
    }
    parser.end_group();

    Ok(())
}

fn parse_format10(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    parser.read::<u16>("Reserved")?;
    parser.read::<u32>("Subtable size")?;
    parse_language32(platform_id, parser)?;
    parser.read::<u32>("First code")?;
    let count = parser.read::<u32>("Number of codes")? as usize;
    parser.read_array::<GlyphId>("Glyph index array", TitleKind::Glyph, count)?;

    Ok(())
}

fn parse_format12(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    parser.read::<u16>("Reserved")?;
    parser.read::<u32>("Subtable size")?;
    parse_language32(platform_id, parser)?;
    let count = parser.read::<u32>("Number of groups")?;

    parser.begin_group_with_value("SequentialMapGroup records", count.to_string());
    for i in 0..count {
        parser.begin_group(format!("Record {}", i));
        parser.read::<u32>("First character code")?;
        parser.read::<u32>("Last character code")?;
        parser.read::<u32>("Starting glyph index")?;
        parser.end_group();
    }
    parser.end_group();

    Ok(())
}

fn parse_format13(platform_id: PlatformId, parser: &mut Parser) -> Result<()> {
    parser.read::<u16>("Reserved")?;
    parser.read::<u32>("Subtable size")?;
    parse_language32(platform_id, parser)?;
    let count = parser.read::<u32>("Number of groups")?;

    parser.begin_group_with_value("ConstantMapGroup records", count.to_string());
    for i in 0..count {
        parser.begin_group(format!("Record {}", i));
        parser.read::<u32>("First character code")?;
        parser.read::<u32>("Last character code")?;
        parser.read::<u32>("Glyph index")?;
        parser.end_group();
    }
    parser.end_group();

    Ok(())
}

fn parse_format14(parser: &mut Parser) -> Result<()> {
    let table_start = parser.offset() - 2;

    parser.read::<u32>("Subtable size")?;
    let count = parser.read::<u32>("Number of records")?;

    struct Record {
        default: bool,
        offset: usize,
    };

    let mut records = Vec::new();
    parser.begin_group_with_value("VariationSelector records", count.to_string());
    for i in 0..count {
        parser.begin_group(format!("Record {}", i));
        parser.read::<U24>("Variation selector")?;
        let def_offset = parser.read::<Offset32>("Offset to Default UVS Table")?.to_usize();
        let non_def_offset = parser.read::<Offset32>("Offset to Non-Default UVS Table")?.to_usize();
        parser.end_group();

        if def_offset != 0 {
            records.push(Record { default: true, offset: table_start + def_offset });
        }

        if non_def_offset != 0 {
            records.push(Record { default: false, offset: table_start + non_def_offset });
        }
    }
    parser.end_group();

    records.sort_by_key(|r| r.offset);
    records.dedup_by_key(|r| r.offset);

    for record in records {
        parser.jump_to(record.offset)?;
        if record.default {
            parser.begin_group("Default UVS table");
            let count = parser.read::<u32>("Number of Unicode character ranges")?;
            for _ in 0..count {
                parser.begin_group("Unicode range");
                parser.read::<U24>("First value in this range")?;
                parser.read::<u8>("Number of additional values")?;
                parser.end_group();
            }
            parser.end_group();
        } else {
            parser.begin_group("Non-Default UVS table");
            let count = parser.read::<u32>("Number of UVS Mappings")?;
            for _ in 0..count {
                parser.begin_group("UVS mapping");
                parser.read::<U24>("Base Unicode value")?;
                parser.read::<GlyphId>("Glyph ID")?;
                parser.end_group();
            }
            parser.end_group();
        }
    }

    Ok(())
}
